GraphicsContextCG is instantiated redundantly to round rects

kkinnunen-apple opened the pull request:
#### efab4b9bc302ea934c09eed6f7fff30a5dfddff5
<pre>
GraphicsContextCG is instantiated redundantly to round rects
<a href="https://bugs.webkit.org/show_bug.cgi?id=262364">https://bugs.webkit.org/show_bug.cgi?id=262364</a>
rdar://116233471

Reviewed by Tim Horton.

GraphicsContext construction does non-trivial work to ensure the context
state is constent. It shouldn&apos;t be called just to round rects.

Extract the rect rounding code to a freestanding function in
CGUtilities.h

Unduplicate interpolation quality enum conversion code to the file, too.

* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::drawPatternCallback):
(WebCore::GraphicsContextCG::roundToDevicePixels const):
(WebCore::cgInterpolationQuality): Deleted.
(WebCore::GraphicsContextCG::scaleForRoundingToDevicePixels const): Deleted.
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cg/PatternCG.cpp:
(WebCore::patternCallback):
* Source/WebCore/platform/ios/wak/WAKView.mm:
(+[WAKView _setInterpolationQuality:]):
(toCGInterpolationQuality): Deleted.

Canonical link: <a href="https://commits.webkit.org/268841@main">https://commits.webkit.org/268841@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53ef00525eebda9ece32e343027bddca430b028c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20848 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21256 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21916 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22738 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19429 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21085 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24495 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21432 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21070 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20839 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18103 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23592 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18003 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18916 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25211 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19083 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/19098 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23120 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19680 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16713 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18921 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4996 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23250 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19495 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/cg/CGUtilities.h:
```diff
@@ -0,0 +1,95 @@
+    auto deviceScaleX = hypot(deviceMatrix.a, deviceMatrix.b);
+    auto deviceScaleY = hypot(deviceMatrix.c, deviceMatrix.d);
+
+    CGPoint deviceOrigin = CGPointMake(rect.x() * deviceScaleX, rect.y() * deviceScaleY);
+    CGPoint deviceLowerRight = CGPointMake((rect.x() + rect.width()) * deviceScaleX,
+        (rect.y() + rect.height()) * deviceScaleY);
```
Why do we have to you `CGPoint` here? The input and the out of this function are `FloatRect`.

kkinnunen-apple replied:
The existing code has that.

shallawa commented on Source/WebCore/platform/graphics/cg/CGUtilities.h:
```diff
@@ -0,0 +1,95 @@
+
+namespace WebCore {
+
+inline CGInterpolationQuality cgInterpolationQuality(InterpolationQuality quality)
```
I think this name is a little bit confusing.  From the first glance, one might think `cgInterpolationQuality()` will return the _current_ interpolation quality. `toCGInterpolationQuality()` is clearer because it indicates it converts something to `CGInterpolationQuality`.

kkinnunen-apple replied:
FWIW, the code style disallows to* functions by default. Though cgInteroplationQuality is not neccessarily in line with the style guide either.

shallawa commented on Source/WebCore/platform/graphics/cg/CGUtilities.h:
```diff
@@ -0,0 +1,95 @@
+
+inline FloatRect cgRoundToDevicePixelsNonIdentity(CGAffineTransform deviceMatrix, FloatRect rect)
+{
+    // It is not enough just to round to pixels in device space. The rotation part of the
```
Should we `ASSERT(!CGAffineTransformIsIdentity(deviceMatrix))`?

kkinnunen-apple replied:
Probably not, the idea probably is that that is the same result..

shallawa commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -1412,57 +1396,17 @@ AffineTransform GraphicsContextCG::getCTM(IncludeDeviceScale includeScale) const
+    CGAffineTransform deviceMatrix;
+    if (!m_userToDeviceTransformKnownToBeIdentity) {
+        deviceMatrix = CGContextGetUserSpaceToDeviceSpaceTransform(contextForState());
+        if (CGAffineTransformIsIdentity(deviceMatrix))
+            m_userToDeviceTransformKnownToBeIdentity = true;
```
I think `m_userToDeviceTransformKnownToBeIdentity` should be `std::optional<bool>` which is set to `std::nullopt` when the transform changes. And here it can be set to 

```
m_userToDeviceTransformIsIdentity = CGAffineTransformIsIdentity(deviceMatrix)
```

So we do not to have to call CGAffineTransformIsIdentity() every time if the transform is not identity.

kkinnunen-apple replied:
Ok, but should we do it in this patch or in some other patch?

shallawa commented on Source/WebCore/platform/graphics/cg/CGUtilities.h:
```diff
@@ -0,0 +1,95 @@
+    return kCGInterpolationDefault;
+}
+
+inline FloatRect cgRoundToDevicePixelsNonIdentity(CGAffineTransform deviceMatrix, FloatRect rect)
```
Does this function have to be CG only? Can't we pass AffineTransform instead and move non CG specific file?

kkinnunen-apple replied:
Yes, but should we do it in this patch or in some other patch?
(I can also rewrite the algorithm to read simpler, but should that be done in this patch or other patch?)

shallawa commented on Source/WebCore/platform/graphics/cg/CGUtilities.h:
```diff
@@ -0,0 +1,95 @@
+    return FloatRect(roundedOrigin, roundedLowerRight - roundedOrigin);
+}
+
+inline FloatRect cgRoundToDevicePixels(CGAffineTransform deviceMatrix, FloatRect rect)
```
Is there a reason to have these functions inline? Can't we have a source and a header files?

kkinnunen-apple replied:
Well, it optimizes pretty much to a*b+c type algorithm, so it's kind of in line with other short inline functions..

kkinnunen-apple committed:
GraphicsContextCG is instantiated redundantly to round rects
https://bugs.webkit.org/show_bug.cgi?id=262364
rdar://116233471

Reviewed by Tim Horton.

GraphicsContext construction does non-trivial work to ensure the context
state is constent. It shouldn't be called just to round rects.

Extract the rect rounding code to a freestanding function in
CGUtilities.h

Unduplicate interpolation quality enum conversion code to the file, too.

* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::drawPatternCallback):
(WebCore::GraphicsContextCG::roundToDevicePixels const):
(WebCore::cgInterpolationQuality): Deleted.
(WebCore::GraphicsContextCG::scaleForRoundingToDevicePixels const): Deleted.
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cg/PatternCG.cpp:
(WebCore::patternCallback):
* Source/WebCore/platform/ios/wak/WAKView.mm:
(+[WAKView _setInterpolationQuality:]):
(toCGInterpolationQuality): Deleted.

Canonical link: https://commits.webkit.org/268841@main
